File: Year 2/Semester 4/MPP/JavaFourthWeek/src/main/java/repository/ShowDBRepository.java
package repository;

import domain.Show;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ShowDBRepository extends AbstractRepository<Integer, Show> {

    public ShowDBRepository(Properties props) {
        super(props);
    }

    private Show getShowFromResultSet(ResultSet resultSet) throws SQLException {
        int id = resultSet.getInt("id");
        String title = resultSet.getString("title");
        LocalDate date = resultSet.getDate("date").toLocalDate();
        String place = resultSet.getString("place");
        int idArtist = resultSet.getInt("id_artist");
        return new Show(id, title, date, place, idArtist);
    }

    @Override
    public Show findById(Integer id) {
        logger.traceEntry("find a show by id");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "select * from show where id = ?")) {
            ps.setInt(1, id);
            try (ResultSet resultSet = ps.executeQuery()) {
                if (resultSet.next()) {
                    Show show = getShowFromResultSet(resultSet);
                    logger.traceExit(show);
                    return show;
                }
            }
        } catch (SQLException exception) {
            logger.error("ERROR for findById in ShowDBRepository: " + exception);
        }
        return null;
    }

    @Override
    public Iterable<Show> getAll() {
        logger.traceEntry("finding all tickets");
        Connection connection = dbUtils.getConnection();
        List<Show> shows = new ArrayList<>();
        try (PreparedStatement ps = connection.prepareStatement("select * from show")) {
            try (ResultSet resultSet = ps.executeQuery()) {
                while (resultSet.next()) {
                    shows.add(getShowFromResultSet(resultSet));
                }
            }
        } catch (SQLException exception) {
            logger.error("ERROR for getAll in ShowDBRepository: " + exception);
        }
        logger.traceExit(shows);
        return shows;
    }

    @Override
    public boolean add(Show show) {
        logger.traceEntry("adding show {}", show);
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "insert into show (title,date,place,id_artist) values (?, ?, ?, ?)")) {
            ps.setString(1, show.getTitle());
            ps.setString(2, show.getDate().toString());
            ps.setString(3, show.getPlace());
            ps.setInt(4, show.getIdArtist());
            int result = ps.executeUpdate();
            logger.trace("saved {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for insert in ShowDBRepository: " + exception);
            return false;
        }
        logger.traceExit("inserted successfully");
        return true;
    }

    @Override
    public boolean delete(Integer id) {
        logger.traceEntry("delete ticket");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("delete from show where id=?")) {
            ps.setInt(1, id);
            int result = ps.executeUpdate();
            logger.trace("deleted {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for delete in ShowDBRepository: " + exception);
            return false;
        }
        logger.traceExit("deleted successfully");
        return true;
    }

    @Override
    public boolean update(Show show) {
        logger.traceEntry("updating ticket");
        Connection connection = dbUtils.getConnection();
        try (PreparedStatement ps = connection.prepareStatement("" +
                "update show set title = ?, date = ?, place = ?, id_artist = ?" +
                " where id = ?")) {
            ps.setString(1, show.getTitle());
            ps.setString(2, show.getDate().toString());
            ps.setString(3, show.getPlace());
            ps.setInt(4, show.getIdArtist());
            ps.setInt(5, show.getId());
            int result = ps.executeUpdate();
            logger.trace("updated {} instance", result);
        } catch (SQLException exception) {
            logger.error("ERROR for update in ShowDBRepository: " + exception);
            return false;
        }
        logger.traceExit("updated successfully");
        return true;
    }
}

File: Year 2/Semester 4/Programming and Projection Enironments/Week 4/C#FourthWeek/SaptamanaAPatra/repository/ShowDBRepository.cs
using log4net;
using SaptamanaAPatra.domain;
using SaptamanaAPatra.repository.interfaces;
using SaptamanaAPatra.utils;
using System.Collections.Generic;
using System.Data.SQLite;

namespace SaptamanaAPatra.repository
{
    internal class ShowDBRepository : IShowRepository
    {
        private static readonly ILog logger = LogManager.GetLogger("ShowDBRepository");

        private IDictionary<string, string> _props;

        public ShowDBRepository(IDictionary<string, string> props)
        {
            this._props = props;
            logger.Info("Creating ShowDBRepository");
        }

        private static Show<int> GetShowFromResultSet(SQLiteDataReader resultSet)
        {
            int id = resultSet.GetInt32(0);
            string title = resultSet.GetString(1);
            string date = resultSet.GetString(2);
            string place = resultSet.GetString(3);
            int idArtist = resultSet.GetInt32(4);
            return new Show<int>(id, title, date, place, idArtist);
        }

        public List<Show<int>> GetAll()
        {
            logger.Info("finding all shows");
            var connection = DBUtils.GetConnection(_props);
            var list = new List<Show<int>>();
            using (var cmd = new SQLiteCommand("select * from show", connection as SQLiteConnection))
            using (var reader = cmd.ExecuteReader())
                while (reader.Read())
                    list.Add(GetShowFromResultSet(reader));
            logger.Info(list);
            return list;
        }


        public Show<int> FindById(int id)
        {
            logger.Info("find a show by id");
            var connection = DBUtils.GetConnection(_props);
            var cmd = new SQLiteCommand("select * from show where id = " + id,
                connection as SQLiteConnection);
            using (var reader = cmd.ExecuteReader())
            {
                if (reader.Read())
                {
                    Show<int> show = GetShowFromResultSet(reader);
                    logger.Info(show);
                    return show;
                }
            }
            return null;
        }

        public bool Add(Show<int> show)
        {
            logger.Info("adding show");
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand
                ("insert into show values (@id, @title, @date, @place, @id_artist)",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", null);
                command.Parameters.AddWithValue("@title", show.Title);
                command.Parameters.AddWithValue("@date", show.Date);
                command.Parameters.AddWithValue("@place", show.Place);
                command.Parameters.AddWithValue("@id_artist", show.IdArtist);
                command.ExecuteNonQuery();
                logger.InfoFormat("inserted successfully {0}", show);
                return true;
            }
        }


        public bool Delete(int id)
        {
            logger.InfoFormat("delete show with id = {0}", id);
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand("delete from show where id = @id",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@id", id);
                command.ExecuteNonQuery();
                logger.InfoFormat("deleted successfully for id {0}", id);
                return true;
            }
        }

        public bool Update(Show<int> show)
        {
            logger.Info("updating show");
            var connection = DBUtils.GetConnection(_props);
            using (var command = new SQLiteCommand("update show set title = @title, " +
                "date = @date, place = @place, id_artist = @id_artist where id = @id",
                      connection as SQLiteConnection))
            {
                command.Parameters.AddWithValue("@title", show.Title);
                command.Parameters.AddWithValue("@date", show.Date);
                command.Parameters.AddWithValue("@place", show.Place);
                command.Parameters.AddWithValue("@id_artist", show.IdArtist);
                command.Parameters.AddWithValue("@id", show.Id);
                command.ExecuteNonQuery();
                logger.Info("updated successfully");
                return true;
            }
        }
    }
}
